src/Compilers/Core/Portable/FileSystem/ICommonCompilerFileSystem.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// Abstraction over the file system that is useful for test hooks
+    /// </summary>
+    internal interface ICommonCompilerFileSystem
+    {
+        bool FileExists(string filePath);
+
+        Stream OpenFile(string filePath, FileMode mode, FileAccess access, FileShare share);
+
+        Stream OpenFileEx(string filePath, FileMode mode, FileAccess access, FileShare share, int bufferSize, FileOptions options, out string normalizedFilePath);
+    }
+
+    internal static class CommonCompilerFileSystemExtensions
+    {
+        /// <summary>
+        /// Open a file and ensure common exception types are wrapped to <see cref="IOException"/>.
+        /// </summary>
+        internal static Stream OpenFileWithNormalizedException(this ICommonCompilerFileSystem fileSystem, string filePath, FileMode fileMode, FileAccess fileAccess, FileShare fileShare)
+        {
+            try
+            {
+                return fileSystem.OpenFile(filePath, fileMode, fileAccess, fileShare);
+            }
+            catch (ArgumentException)
+            {
+                throw;
+            }
+            catch (DirectoryNotFoundException e)
+            {
+                throw new FileNotFoundException(e.Message, filePath, e);
```

Comment:
>throw new FileNotFoundException(e.Message, filePath, e) [](start = 16, length = 55)

Why throw a `FileNotFoundException` rather than the original `DirectoryNotFoundException`?

Reply:
This is just the existing code from `FileUtilities` that I copied here. Not changing the semantics, maintaining them for now but have a follow up item to revisit this logic. 

https://github.com/dotnet/roslyn/issues/52241